KAFKA-3179 Fix seek on compressed messages

becketqin opened the pull request:
The fix itself is simple. 

Some explanation on unit tests. Currently we the vast majority of unit test is running with uncompressed messages.  I was initially thinking about run all the tests using compressed messages. But it seems uncompressed messages are necessary in a many test cases because we need the bytes sent and appended to the log to be predictable. In most of other cases, it does not matter whether the message is compressed or not, and compression will slow down the unit test. So I just added one method in the BaseConsumerTest to send compressed messages whenever we need it.

becketqin committed:
KAFKA-3179 Fix seek on compressed messages

ijuma commented:
cc @hachikuji

ijuma commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -559,6 +559,9 @@ private void handleFetchResponse(ClientResponse resp, FetchRequest request) {
                 for (LogEntry logEntry : records) {
+                    // Skip the messages earlier than current position.
+                    if (logEntry.offset() < position)
+                        continue;
```
Nitpick: it would be more composable to do:

``` java
if (logEntry.offset() >= position) {
...
}
```

The reason is that we can freely move the if block to another method (if it needs to be reused or becomes more complex) and the code still works correctly. `continue` and `break` couple code to the surrounding loop (unnecessarily in many cases).

ijuma commented on core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala:
```diff
@@ -276,7 +281,20 @@ abstract class BaseConsumerTest extends IntegrationTestHarness with Logging {
+      retries = 0, lingerMs = Long.MaxValue, props = Some(producerProps))
+    (0 until numRecords).foreach { i =>
+      producer.send(new ProducerRecord(tp.topic(), tp.partition(), s"key $i".getBytes, s"value $i".getBytes))
+    }
```
Somewhat of a nitpick too, but we could extract a method for the following since it's duplicated in both `send*Records` methods.

``` scala
    (0 until numRecords).foreach { i =>
       producer.send(new ProducerRecord(tp.topic(), tp.partition(), s"key $i".getBytes, s"value $i".getBytes))
     }
```

becketqin committed:
Addressed Ismael's comments

jjkoshy commented on core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala:
```diff
@@ -274,9 +279,24 @@ abstract class BaseConsumerTest extends IntegrationTestHarness with Logging {
+    this.producers(0).flush()
+  }
+
+  protected def sendCompressedRecords(numRecords: Int, tp: TopicPartition) {
```
Can we make this `sendRecords` and have it take a compression-codec argument instead of hard-coding to gzip?

jjkoshy commented on core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:
```diff
@@ -266,7 +266,7 @@ class PlaintextConsumerTest extends BaseConsumerTest {
     val consumer = this.consumers(0)
     val totalRecords = 50L
-    sendRecords(totalRecords.toInt)
+    sendCompressedRecords(totalRecords.toInt, tp)
```
We should probably also test with uncompressed messages - if you make the above change to `sendRecords` you can either make the compression-codec a parameterized argument of the test or just send both compressed and uncompressed in a single run.

jjkoshy commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetcherTest.java:
```diff
@@ -464,8 +464,15 @@ public void testQuotaMetrics() throws Exception {
+            // We need to make sure the message offset grows.
+            if (i > 1) {
+                this.records = MemoryRecords.emptyRecords(ByteBuffer.allocate(1024), CompressionType.NONE);
+                this.records.append((long) i * 3, "key".getBytes(), "value-1".getBytes());
```
value-0, value-1, value-2

becketqin committed:
Addressed Joel's comments

becketqin committed:
Addressed Joel's comments.

becketqin commented:
@jjkoshy I just addressed your comments. Would you take another look? It seems parameterized test  can only apply to class level, I just added a separate test method. Thanks.

jjkoshy commented on core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:
```diff
@@ -263,10 +263,31 @@ class PlaintextConsumerTest extends BaseConsumerTest {
     val totalRecords = 50L
-    sendRecords(totalRecords.toInt)
+    sendRecords(totalRecords.toInt, tp)
+    consumer.assign(List(tp).asJava)
```
Rather than duplicate this code I think there are a couple of ways you can cover both cases in the same method fairly cleanly.

jjkoshy commented on core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala:
```diff
@@ -274,9 +279,28 @@ abstract class BaseConsumerTest extends IntegrationTestHarness with Logging {
+    sendRecords(numRecords, tp, None)
+  }
+
+  protected def sendRecords(numRecords: Int, tp: TopicPartition, codec: Option[String] = None) {
```
After a second look I'm not a big fan of this change. Basically, the `IntegrationTestHarness` is ill-equipped to handle multiple types of producers. These changes make the API a bit misleading. For e.g., if the producer-config that you supply to the harness is configured with compression turned on then the preceding `sendRecords` API would send compressed records and this would be redundant. While this sounds out of scope for this jira I'm wondering if it is worth thinking of a simple way to improve this.

jjkoshy commented on core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala:
```diff
@@ -15,15 +15,18 @@ package kafka.api
+import java.util.{Properties, ArrayList}
 import org.junit.Assert._
+import org.junit.runner.RunWith
+import org.junit.runners.Parameterized
```
Unused imports

jjkoshy commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetcherTest.java:
```diff
@@ -464,8 +464,15 @@ public void testQuotaMetrics() throws Exception {
+                this.records = MemoryRecords.emptyRecords(ByteBuffer.allocate(1024), CompressionType.NONE);
+                this.records.append((long) i * 3, "key".getBytes(), "value-0".getBytes());
+                this.records.append((long) i * 3 + 1, "key".getBytes(), "value-1".getBytes());
+                this.records.append((long) i * 3 + 2, "key".getBytes(), "value-2".getBytes());
```
Maybe:

```
for (int v = 0; v < 3; v++) {
    this.records.append((long) i * 3 + v, "key".getBytes(), String.format("value-%d", v).getBytes());
}
```

becketqin committed:
Addressed Joel's comments.

becketqin commented:
@jjkoshy Thanks for the comments. I moved the sendCompressedMessage to PlaintextConsumerTest to avoid polluting BaseConsumerTest. It looks we only need sending compressed message there.

becketqin commented:
Test failure is irrelevant.

auradkar commented on core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:
```diff
@@ -266,21 +268,37 @@ class PlaintextConsumerTest extends BaseConsumerTest {
-    assertEquals(mid, consumer.position(tp))
-    consumeAndVerifyRecords(consumer, numRecords = 1, startingOffset = mid.toInt, startingKeyAndValueIndex = mid.toInt)
+    // Test seek non-compressed message
+    val quarter = totalRecords / 2
```
may want to rename these variables. quarter and threeQuarters are slightly confusing

auradkar commented on core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:
```diff
@@ -266,21 +268,37 @@ class PlaintextConsumerTest extends BaseConsumerTest {
+    producerProps.setProperty(ProducerConfig.LINGER_MS_CONFIG, Long.MaxValue.toString)
+    val producer = TestUtils.createNewProducer(brokerList, securityProtocol = securityProtocol, trustStoreFile = trustStoreFile,
+        retries = 0, lingerMs = Long.MaxValue, props = Some(producerProps))
+    sendRecords(producer, numRecords, tp, startingKeyAndValueIndex)
```
shouldn't we shut this down?

jjkoshy replied:
rather than change `sendRecords` can we just have two cases in the test (compressed/uncompressed) that send to different topic-partitions?

auradkar commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetcherTest.java:
```diff
@@ -464,8 +464,15 @@ public void testQuotaMetrics() throws Exception {
 
         // normal fetch
         for (int i = 1; i < 4; i++) {
+            // We need to make sure the message offset grows.
```
the comment doesn't make it clear why we need to make sure the message offset grows

becketqin committed:
Addressed Adi's comments

becketqin committed:
Changed the test code again.

jjkoshy commented:
+1 - thanks Becket

ijuma commented:
@jjkoshy, we probably want to double-commit this to 0.9.0 as well.

jjkoshy commented:
true - will cherry-pick that
